File: app/src/main/java/me/zhanghai/android/files/compat/ErrnoExceptionCompat.java
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.compat;

import android.system.ErrnoException;

import androidx.annotation.NonNull;
import me.zhanghai.android.files.reflected.ReflectedAccessor;
import me.zhanghai.android.files.reflected.ReflectedField;
import me.zhanghai.android.files.reflected.RestrictedHiddenApi;

public class ErrnoExceptionCompat {

    static {
        ReflectedAccessor.allowRestrictedHiddenApiAccess();
    }

    @RestrictedHiddenApi
    private static final ReflectedField<ErrnoException> sFunctionNameField = new ReflectedField<>(
            ErrnoException.class, "functionName");

    private ErrnoExceptionCompat() {}

    @NonNull
    public static String getFunctionName(@NonNull ErrnoException errnoException) {
        return sFunctionNameField.getObject(errnoException);
    }
}

File: app/src/main/java/me/zhanghai/android/files/compat/ErrnoExceptionCompat.kt
/*
 * Copyright (c) 2018 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.compat

import android.system.ErrnoException
import me.zhanghai.java.reflected.ReflectedField

@RestrictedHiddenApi
private val functionNameField = ReflectedField(ErrnoException::class.java, "functionName")

val ErrnoException.functionNameCompat: String
    get() = functionNameField.getObject(this)
